Refactor `ThreadList::WillResume()` to prepare to support reverse execution

rocallahan opened the pull request:
These changes are designed to not change any behavior, but to make it easy to add code to choose the direction of execution after we've identified which thread(s) to run but before we add any `ThreadPlanStepOverBreakpoint`s. And honestly I think they make the existing code a bit clearer.

rocallahan committed:
[lldb] Move thread-notification setup to happen later

This lets us check `run_me_only_list.GetSize()` instead of `wants_solo_run`.

rocallahan committed:
[lldb] Make `Thread::WillResume()` report whether it pushed a `ThreadPlanStepOverBreakpoint`

We'll need this later to determine whether we need to account for
the newly-pushed `ThreadPlanStepOverBreakpoint`.

rocallahan committed:
[lldb] Use `thread_sp` in more places in `ThreadList::WillResume()`

This makes the code prettier and more consistent.

rocallahan committed:
[lldb] Move `SetupForResume()` calls to happen after the first pass populating `run_me_only_list`

Eventually we will need to determine the execution direction after populating
`run_me_only_list` and before calling `SetupForResume()`.

rocallahan committed:
[lldb] Set `wants_solo_run` based on whether `run_me_only_list` is empty

We no longer need a dedicated loop to set `wants_solo_run`.

rocallahan committed:
[lldb] Move `thread_to_run` selection to happen before `SetupForResume()` is called

Later, we'll need to know which thread will run before we determine the
execution direction, which has to happen before we call `SetupForResume()`.
We fix up `thread_to_run` after calling `SetupForResume()`, if required.

rocallahan committed:
[lldb] Replace `stop_others_thread_sp` with `thread_to_run`

We don't need `stop_others_thread_sp` to be a separate variable anymore,
we can just use `thread_to_run` from the start.

jimingham commented on lldb/source/Target/ThreadList.cpp:
```diff
@@ -582,31 +531,92 @@ bool ThreadList::WillResume() {
+      if (thread_sp->SetupForResume()) {
+        // You can't say "stop others" and also want yourself to be suspended.
+        assert(thread_sp->GetCurrentPlan()->RunState() != eStateSuspended);
+        run_me_only_list.AddThread(thread_sp);
```
Why does this do any good?  You didn't clear the list, and you aren't going to add threads here that didn't get added in the first loop.  So you're just adding them twice.

jimingham replied:
And anyway, if wants_solo_run is false, you end up adding every thread that needs to step over a breakpoint to the run_me_only list, but then don't do anything with any of them.

jimingham commented on lldb/source/Target/ThreadList.cpp:
```diff
@@ -582,31 +531,92 @@ bool ThreadList::WillResume() {
+    // negotiation, but they're coming in after the fact, and the threads that
+    // are already set up should take priority.
+    if (thread_sp->GetResumeState() != eStateSuspended &&
+        (!wants_solo_run || (*pos)->GetCurrentPlan()->StopOthers())) {
```
*pos -> thread_sp

rocallahan replied:
Thanks, done.

jimingham commented on lldb/source/Target/ThreadList.cpp:
```diff
@@ -582,31 +531,92 @@ bool ThreadList::WillResume() {
+  // Don't set to resume suspended threads, and if any thread wanted to stop
+  // others, only call setup on the threads that request StopOthers...
+  bool wants_solo_run = run_me_only_list.GetSize(false) > 0;
+  for (pos = m_threads.begin(); pos != end; ++pos) {
```
I must admit I find this version harder to reason about than the original.  Particularly, I find this second loop confusing.  You have already figured out whether you are going to run solo and which thread you are going to run, by the time you get to this point.  Then it looks like you do that computation a second time in a slightly different way, overwriting the decisions you made above in a fairly head-scratching way.

I would have expected at this point that if `thread_to_run` is not empty, you could just call SetupForResume on that one thread, and then just that thread will run.  It's not at all clear to me what the purpose of redoing the computation is.

There's one difference between the two computations.  In the first loop, you give every thread that wants to run solo a chance to be the one to do that and pick among them equally.  Then in this loop if the thread you chose to run doesn't return `true` from `SetupForResume` (because it isn't at a breakpoint) but there's another wants to run solo thread that is at a breakpoint, we switch to that one.

You are changing behavior here because you prioritize threads that need to step over breakpoints in a way that we didn't before, but the logic seems unclear to me.

Can you say in words what behaviors you are intending to impose with this extra "am I stepping over a breakpoint" check?

rocallahan replied:
> I would have expected at this point that if thread_to_run is not empty, you could just call SetupForResume on that one thread, and then just that thread will run.

Yes, that's a good idea. I've done that in the new revision. It definitely helps.

> You are changing behavior here because you prioritize threads that need to step over breakpoints in a way that we didn't before, but the logic seems unclear to me.

Putting the `SetupForResume()` loop behind a check that `thread_to_run` is non-empty addresses this.

> Can you say in words what behaviors you are intending to impose with this extra "am I stepping over a breakpoint" check?

The goal is to decide which thread to run before we call `SetupForResume()`. I'm trying to change behavior as little as possible while achieving that.

Note that with the new code, there are only at most two loops over all threads, whereas the existing code before this PR always does three. And now, if the first loop identifies a thread to run because some thread returns true for `StopOthers()`, we don't need the second loop.

rocallahan replied:
FYI the new code does change behavior --- when some threads return true for `StopOthers()`, the old code calls `SetupForResume()` on all threads that return true for `StopOthers()`, but now we only call `SetupForResume()` on the one thread we have chosen to run.

Also, when no threads return true for `StopOthers()` but multiple threads have breakpoints that need to be skipped, the old code would choose randomly between those threads, but the new code picks the last one we encounter looping over `m_threads` (if none return true for `ShouldRunBeforePublicStop()`).

rocallahan committed:
[lldb] Replace check of `run_me_only_list.GetSize()` with test of `thread_to_run` against null

This is clearer and it means we don't have to update `run_me_only_list` when `SetupForResume()`
returns true, which was confusing.

jimingham reviewed (approved):
That looks better.  The one functional change - not telling a bunch of threads to set themselves up for running when we know they aren't going to get to run - is better behavior, and I can't see how it would cause problems.

rocallahan commented:
Great, thanks! Hopefully someone can merge this?

jimingham commented:
Done.

rocallahan commented:
Thanks!!
